util/project_allocation_mapper.py

```diff
@@ -0,0 +1,509 @@
+        if self.is_from_db:
+            keycloak_client = KeycloakClient()
+            for proj in keycloak_client.get_user_projects_by_username(username):
+                project = portal_proj.objects.filter(charge_code=proj)
+                if len(project) == 0:
+                    logger.error('Couldn\'t find project {} in portal'.format(proj))
+                else:
+                    project = self.portal_to_tas_proj_obj(project[0], to_pytas_model=to_pytas_model)
+                    project_allocs = self._get_project_allocations(project, to_pytas_model=to_pytas_model)
+                    self.set_attr(project, 'allocations', project_allocs)
+                    user_projects.append(project)
+        else:
+            if to_pytas_model:
+                user_projects = tas_proj.list(username=username)
+            else:
+                user_projects = self.tas.projects_for_user(username=username)
+                        
+            for proj in user_projects:
+                try:
+                    extras = ProjectExtras.objects.get(tas_project_id=self.get_attr(proj, 'id'))
+                    try:
+                        proj.__dict__['nickname'] = extras.nickname
+                    except:
+                        proj['nickname'] = extras.nickname
+                except ProjectExtras.DoesNotExist:
+                    project_nickname = None
+            
+        return user_projects
+    
+    def get_project_members(self, tas_project):
+        users = []
+        if self.is_from_db:
+            keycloak_client = KeycloakClient()
+            pi_username = tas_project.pi.username
+            for username in keycloak_client.get_project_members_by_charge_code(tas_project.chargeCode):
+                # assign role
+                role = 'Standard'
+                if username == pi_username:
+                    role = 'PI'
+                users.append(self.portal_keycloak_user_to_tas_obj(username, to_pytas_model=True, role=role))
+        else:
+            users = tas_project.get_users()
+        return users
+    
+    def get_project(self, project_id):
+        if self.is_from_db:
+            try:
+                project = portal_proj.objects.get(pk=project_id)
+            except portal_proj.DoesNotExist:
+                return None
+            project = self.portal_to_tas_proj_obj(project, to_pytas_model=True)
+            project_allocs = self._get_project_allocations(project)
+            self.set_attr(project, 'allocations', project_allocs)
+            return project
+        else:
+            tas_project = tas_proj(project_id)
+            try:
+                extras = ProjectExtras.objects.get(tas_project_id=tas_project.id)
+                tas_project.__dict__['nickname'] = extras.nickname
+            except ProjectExtras.DoesNotExist:
+                project_nickname = None
+            return tas_project
+
+    def allocation_approval(self, data, host):
+        if self.is_from_db:
```

Comment:
The code choice here depends on if the _project_ is federated, not the user. Similarly, try to fetch it from portal first--if it is found, you know it's in the "DB" case.